source/dealii/dealii_matrix_free_hierarchy_helpers.cc

```diff
@@ -69,15 +72,157 @@ DealIIMatrixFreeHierarchyHelpers<dim, VectorType>::build_restrictor(
+                      eigenvalues[row];
+            }
+          }
+          else
+          {
+            for (unsigned int k = 0; k < n_elem; ++k)
+            {
+              delta_eig[k] =
+                  delta_eigenvector_matrix->el(row, dof_indices_map[k]);
+            }
+          }
+
+          // Perform the matrix-vector multiplication
+          dealii::Vector<ScalarType> correction(n_elem);
+          dealii_mesh_evaluator->matrix_free_evaluate_agglomerate(
+              agglomerate_dof_handler, delta_eig, correction);
+
+          // We would like to fill the delta correction matrix but we can't
+          // because we don't know the sparsity pattern. So we accumulate all
+          // the values and then fill the matrix using the set() function.
+          for (unsigned int k = 0; k < n_elem; ++k)
+            delta_correction_acc[std::make_pair(row, dof_indices_map[k])] +=
+                correction[k];
+        }
+      }
+
+      is_halo_agglomerate = true;
+    }
+
+    // Fill delta_correction_matrix
+    for (auto const &entry : delta_correction_acc)
+      delta_correction_matrix.set(entry.first.first, entry.first.second,
+                                  entry.second);
+    delta_correction_matrix.compress(dealii::VectorOperation::insert);
+
+    // Add the eigenvector matrix and the delta correction matrix to create ap
+    Epetra_CrsMatrix *ap = nullptr;
+    // We want to use functions that have been deprecated in deal.II but they
+    // won't be removed in the foreseeable future
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+    Epetra_Map range_map = eigenvector_matrix->domain_partitioner();
+    Epetra_Map domain_map = eigenvector_matrix->range_partitioner();
+#pragma GCC diagnostic pop
+
+    bool const transpose = true;
+    int error_code = EpetraExt::MatrixMatrix::Add(
+        eigenvector_matrix->trilinos_matrix(), transpose, 1.,
+        delta_correction_matrix.trilinos_matrix(), transpose, 1., ap);
+    ap->FillComplete(domain_map, range_map);
+    ASSERT(error_code == 0, "Problem when adding matrices");
+
+    // Copy the Epetra_CrsMatrix to a dealii::TrilinosWrappers::SparseMatrix
+    auto dealii_ap = std::make_shared<dealii::TrilinosWrappers::SparseMatrix>();
+    dealii_ap->reinit(*ap);
+    delete ap;
+
+    _ap_operator.reset(new DealIITrilinosMatrixOperator<VectorType>(dealii_ap));
+  }
+  else
+  {
+    AMGe_host<dim, DealIIMatrixFreeMeshEvaluator<dim>, VectorType> amge(
+        comm, dealii_mesh_evaluator->get_dof_handler(), eigensolver_params);
+
+    amge.setup_restrictor(agglomerate_params, n_eigenvectors, tolerance,
+                          *dealii_mesh_evaluator, locally_relevant_global_diag,
+                          *restrictor_matrix);
+  }
 
   std::shared_ptr<Operator<VectorType>> op(
       new DealIITrilinosMatrixOperator<VectorType>(restrictor_matrix));
 
   return op;
-}
+} // namespace mfmg
```

Comment:
This comment shouldn't be there or there is a problem because the functions below should still be in the namespace